scripts/vehicle_node.py

```diff
@@ -201,78 +203,115 @@ def update_position(self, event):
-            joint_velocity_data = copy.deepcopy(self.joint_twist)
-
-        #time elapsed since last update position call
-        if hasattr(event, 'last_real'):
-            if event.last_real is None:
-                time = rospy.Duration(0.05)
-            else:
-                time = event.current_real - event.last_real
-        
-        time = time.to_sec()
-
-        #Calculate angle turned in the given time using omega = theta/time
-        angle = joint_velocity_data.angular.y*time
-
-        #Calculate distance travelled in the given time using linear velocity = arc distance/time
-        distance = joint_velocity_data.linear.z*time
-
-        #Calculate pitch of the robot joint
-        self.joint_pitch += angle
-
-        #Calculate vehicle x, y, z position coordinates
-        self.joint_pose.position.z += (distance)*cos(self.joint_pitch)
+            joint_data = copy.deepcopy(self.joint)
 
-        br = tf.TransformBroadcaster()
-        q2 = tf.transformations.quaternion_from_euler(0.0, self.joint_pitch, -3.14159)
-        translation =  [-0.15, 0.0, 0.0]
-
-        br.sendTransform((-0.15, 0.0, 0.0), 
-                        q2, 
-                        rospy.Time.now(), 
-                        "joint_frame", 
-                        "vehicle_frame")
+        #If the data is empty on first run, fill with 0.0
+        if not joint_data.velocity:
+            joint_data.velocity = [0.0,0.0]
         
-        vehicle_t_joint1 = tf.transformations.translation_matrix(translation)
-        vehicle_R_joint1   = tf.transformations.quaternion_matrix(q2)
-        vehicle_T_joint1 = np.zeros((4,4))
-        vehicle_T_joint1[0:3,0:3] = vehicle_R_joint1[0:3,0:3]
-        vehicle_T_joint1[:4,3] = vehicle_t_joint1[:4,3]
-        # print vehicle_T_joint1
-
-        footprint = np.array([[0.0],[0.0],[0.0],[1.0]])
-        footprint = np.matmul(vehicle_T_joint1, footprint)
-
-        self.joint_pose.position.x = footprint[0,0]
-        self.joint_pose.position.y = footprint[1,0]
-        self.joint_pose.position.z = footprint[2]
-        self.joint_pose.orientation.x = q2[0]
-        self.joint_pose.orientation.y = q2[1]
-        self.joint_pose.orientation.z = q2[2]
-        self.joint_pose.orientation.w = q2[3]
+        #Calculate angle based on velocity data and time
+        angle = joint_data.velocity[0]*time
+        angle2 = joint_data.velocity[1]*time
+
+        self.joint1_pitch += angle
+        self.joint2_pitch += angle2
+
+        #Transformations from vehicle frame to Joint1 and Joint2
+    
+        #Static rotation about z-axis 
+        static_rot = tf.transformations.quaternion_from_euler(0.0, 0.0, 3.14159)
+        translation =  [0.0, 0.0, 0.0]
+        V_T_SRz = tf.transformations.quaternion_matrix(static_rot)
+        V_T_SRz[:3,3] = np.array(translation)
+
+
+        #Dynamic rotation of Joint 1
+        dynamic_rot = tf.transformations.quaternion_from_euler(0.0, self.joint1_pitch, 0.0)
```

Comment:
I agree that it's possible to use this helper function, but it hides whether or not you understand the underlying transform math.

Could you switch this to directly generate the 4x4 transform `SRz_T_J1`, directly using trigonometry and your understanding of reference frames?